qcfractal/storage_sockets/sqlalchemy_socket.py

```diff
@@ -1897,33 +1897,53 @@ def queue_submit(self, data: List[TaskRecord]):
 
         meta = add_metadata_template()
 
-        results = []
+        results = ["placeholder"] * len(data)
+
         with self.session_scope() as session:
+            # preserving all the base results for later check
+            all_base_results = [record.base_result for record in data]
+            query_res = (
+                session.query(TaskQueueORM.id, TaskQueueORM.base_result_id)
+                .filter(TaskQueueORM.base_result_id.in_(all_base_results))
+                .all()
+            )
+
+            # constructing a dict of found tasks and their ids
+            found_dict = {str(base_result_id): str(task_id) for task_id, base_result_id in query_res}
+            new_tasks, new_idx = [], []
+            duplicate_idx = []
             for task_num, record in enumerate(data):
-                try:
+
+                if found_dict.get(record.base_result):
+                    # if found, get id from found_dict
+                    # Note: found_dict may return a task object because the duplicate id is of an object in the input.
+                    results[task_num] = found_dict.get(record.base_result)
+                    # add index of duplicates
+                    duplicate_idx.append(task_num)
+                    meta["duplicates"].append(task_num)
+
+                else:
                     task_dict = record.dict(exclude={"id"})
-                    # # for compatibility with mongoengine
-                    # if isinstance(task_dict['base_result'], dict):
-                    #     task_dict['base_result'] = task_dict['base_result']['id']
                     task = TaskQueueORM(**task_dict)
-                    task.priority = task.priority.value  # Must be an integer for sorting
-                    session.add(task)
-                    session.commit()
-                    results.append(str(task.id))
-                    meta["n_inserted"] += 1
-                except IntegrityError as err:  # rare case
-                    # print(str(err))
-                    session.rollback()
-                    # TODO: merge hooks
-                    task = session.query(TaskQueueORM).filter_by(base_result_id=record.base_result).first()
-                    self.logger.warning("queue_submit got a duplicate task: {}".format(task.to_dict()))
-                    results.append(str(task.id))
-                    meta["duplicates"].append(task_num)
-                # except Exception as err:
-                #     self.logger.warning('queue_submit submission error: {}'.format(str(err)))
-                #     meta["success"] = False
-                #     meta["errors"].append(str(err))
-                #     results.append(None)
+                    new_idx.append(task_num)
+                    task.priority = task.priority.value
```

Comment:
Add the comment
# Must be an integer for sorting

Reply:
@codtiger

Reply:
Done in the new commit